docker/elasticsearch-setup/create-indices.sh

```diff
@@ -14,22 +14,102 @@ else
+    return 1
+}
+
+function reindex() {
+  source_index=$1
+  target_index="$1_$(date +%s)"
+
+  #create target index with latest index config
+  curl -XPUT "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$target_index" -H 'Content-Type: application/json' --data @/tmp/data
+
+  #reindex the documents in source index to target index.
+  # One of the assumption here is that we only add properties to document when index-config is evolved.
+  # In case a property is deleted from document, it will still be reindexed in target index as default behaviour and
+  # it is not breaking the code. If still needs to be purged from target index, use "removed" property in POST data.
+  curl -XPOST "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/_reindex?pretty" -H 'Content-Type: application/json' \
+    -d "{\"source\":{\"index\":\"$source_index\"},\"dest\":{\"index\":\"$target_index\"}}"
+
+  if check_reindex "$source_index" "$target_index"
+  then
+    #checking if source index is concrete index or alias
+    if [ $(curl -o /dev/null -s -w "%{http_code}" "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/_alias/$source_index") -eq 404 ]
+    then
+      curl -XDELETE "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$source_index"
+    else
+      concrete_index_name=$(curl -XGET "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/_alias/$source_index" | jq 'keys[]' | head -1 | tr -d \")
+      curl -XDELETE "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$concrete_index_name"
+    fi
+
+    curl -XPOST "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/_aliases" -H 'Content-Type: application/json' \
+      -d "{\"actions\":[{\"remove\":{\"index\":\"*\",\"alias\":\"$source_index\"}},{\"add\":{\"index\":\"$target_index\",\"alias\":\"$source_index\"}}]}"
+
+    echo -e "\nReindexing to $target_index is complete"
+    return 0
+  else
+    echo -e "\nReindexing to $target_index is failure"
+    return 1
+  fi
+}
+
+function create_index() {
+  generate_index_file "index/$2" "index/$3" /tmp/data
+
+  #checking if index(or alias) exists
+  if [ $(curl -o /dev/null -s -w "%{http_code}" "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$1") -eq 404 ]
+  then
+    echo -e '\ncreating index' "$1"
+
+    curl -XPUT "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$1" -H 'Content-Type: application/json' --data @/tmp/data
+    return 0
+  else
+    echo -e '\ncomparing with existing version of index' "$1"
+
+    curl -XGET "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$1/_settings" | jq '.. | .settings? | select(. != null)' | \
+      jq '. | {"index":{"max_ngram_diff": .index.max_ngram_diff, "analysis": .index.analysis}}' > /tmp/existing_setting
```

Comment:
What is this last jq step for?

Reply:
ElasticSearch's '_settings' endpoint returns some more attributes which are not configured by us. For example, in below output
```
{
  "dataprocessdocument_1617072295": {
    "settings": {
      "index": {
        "max_ngram_diff": "47",
        "number_of_shards": "1",
        "provided_name": "dataprocessdocument_1617072295",
        "creation_date": "1617072295842",
        "analysis": {
          "filter": {
            "autocomplete_filter": {
              "type": "edge_ngram",
.......
```
attributes like 'number_of_shards','provided_name','creation_date', etc are not configured by us. Thus, the last jq step is to filter out such attributes and to only compare what we intend to configure.

Reply:
In the new commit, i have done the automated discovery of expected keys in settings.json, instead of hardcoding the filter `jq '. | {"index":{"max_ngram_diff": .index.max_ngram_diff, "analysis": .index.analysis}}'`.